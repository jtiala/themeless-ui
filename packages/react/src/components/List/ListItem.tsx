import { CommonComponentProps } from "@themeless-ui/utils";
import { ReactNode } from "react";

export type ListItemProps = {
  /**
   * List item content.
   */
  children?: ReactNode;
} & CommonComponentProps;

/**
 * Display a text ListItem.
 */
export function ListItem({ children, testId }: ListItemProps) {
  return <li data-testid={testId}>{children}</li>;
}
