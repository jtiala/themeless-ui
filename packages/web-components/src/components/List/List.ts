/* eslint-disable lit/no-invalid-html */
/* eslint-disable lit/binding-positions */
import style from "@themeless-ui/style/dist/list.css?inline";
import { cn } from "@themeless-ui/utils";
import { nothing, unsafeCSS } from "lit";
import { customElement, property } from "lit/decorators.js";
import { html, unsafeStatic } from "lit/static-html.js";
import { TUIComponent } from "../TUIComponent";

type ListType = "ordered" | "unordered";

const className = cn("list");

@customElement("tui-list")
export class List extends TUIComponent {
  /**
   * Type of the list.
   *
   * @default unordered
   */
  @property()
  type?: ListType;

  static styles = unsafeCSS(style);

  override render() {
    const tag = unsafeStatic(this.type === "ordered" ? "ol" : "ul");

    return html`
      <${tag}
        class="${className}"
        id="${this.id || nothing}"
        data-testid="${this.testId || nothing}"
      >
        <slot></slot>
      </${tag}>
    `;
  }
}

declare global {
  interface HTMLElementTagNameMap {
    "tui-list": List;
  }
}
