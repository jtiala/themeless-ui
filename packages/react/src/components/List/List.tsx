import { CommonComponentProps, cn } from "@themeless-ui/utils";
import { ReactElement, createElement } from "react";
import { ListItem, ListItemProps } from "./ListItem";

type ListType = "ordered" | "unordered";
type ListItemElement = ReactElement<ListItemProps, typeof ListItem>;

export type ListProps = {
  /**
   * Type of the list.
   *
   * @default unordered
   */
  type?: ListType;

  /**
   * One or more <List.Item> elements.
   */
  children?: ListItemElement | ListItemElement[];
} & CommonComponentProps;

const className = cn("list");

/**
 * Display a list of items.
 */
export function List({ type, children, testId }: ListProps) {
  const element = type === "ordered" ? "ol" : "ul";

  return createElement(element, { className, "data-testid": testId }, children);
}

List.Item = ListItem;
