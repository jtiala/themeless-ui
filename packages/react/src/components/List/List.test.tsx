import {
  componentToJson,
  describe,
  expect,
  it,
  render,
  renderer,
  screen,
} from "../../../test";
import { List } from "./List";

const items = [
  <List.Item key="1">Lorem ipsum</List.Item>,
  <List.Item key="2">Dolor sit amet</List.Item>,
  <List.Item key="3">Consectetur adipiscing elit</List.Item>,
];

const unorderedList = <List testId="list">{items}</List>;
const orderedList = (
  <List type="ordered" testId="list">
    {items}
  </List>
);

describe("List", async () => {
  it("should render unordered list", () => {
    expect(componentToJson(renderer.create(unorderedList))).toMatchSnapshot();

    render(unorderedList);

    const list = screen.getByTestId("list");
    const listItems = list.querySelectorAll("li");

    expect(list).toBeInTheDocument();
    expect(list.tagName.toLowerCase()).toBe("ul");
    expect(listItems.length).toBe(3);
    expect(listItems[0]).toHaveTextContent("Lorem ipsum");
    expect(listItems[1]).toHaveTextContent("Dolor sit amet");
    expect(listItems[2]).toHaveTextContent("Consectetur adipiscing elit");
  });

  it("should render ordered list", () => {
    expect(componentToJson(renderer.create(orderedList))).toMatchSnapshot();

    render(orderedList);

    const list = screen.getByTestId("list");

    expect(list).toBeInTheDocument();
    expect(list.tagName.toLowerCase()).toBe("ol");
  });
});
