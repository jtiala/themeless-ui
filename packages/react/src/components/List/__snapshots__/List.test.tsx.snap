// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`List > should render ordered list 1`] = `
<ol
  className="tui__list"
  data-testid="list"
>
  <li>
    Lorem ipsum
  </li>
  <li>
    Dolor sit amet
  </li>
  <li>
    Consectetur adipiscing elit
  </li>
</ol>
`;

exports[`List > should render unordered list 1`] = `
<ul
  className="tui__list"
  data-testid="list"
>
  <li>
    Lorem ipsum
  </li>
  <li>
    Dolor sit amet
  </li>
  <li>
    Consectetur adipiscing elit
  </li>
</ul>
`;
